### PR TITLE
Upgrade VSIX V2->V3 - Closes #4196

### LINKED ISSUE #4196: migrate vsix from v2 vsix to v3 vsix
Tracking via internal bug 384559

### PR DESCRIPTION
**Pull request by nkolev92**
This closes NuGet/Home#4196

As of D15, VS Team plans to drop support for VSIX V2.
V3 VSIX adds some new requirements to both csproj and the manifest.

V3 also dropped supports for ShortcutPath, that MSBuild uses. 
Due to that we have to install NuGet in a static location in CommonExtensions/Microsoft/NuGet.

By doing that though we lose the ability to upgrade our VSIX easily. We cannot simply uninstall an AllUsers VSIX because it'll break workflows that depend on NuGet. Due to that they don't support uninstalling of VSIXs. 

To alleviate this, we will build another VSIX that's experimental.
This allows it to be installed a user's location. However the drawback of this is that MSBuild will still use the CommonExtensions/Microsoft/NuGet version. There's currently no way to support both scenarios.

To minimize changes to our current workflows, the 
/artifacts/VS15/NuGet.Tools.vsix version is an experimental. That it meant to be used in private build and will be the one that's wrapped in the 1ES package.
/artifacts/VS15/Insertable/NuGet.Tools.vsix is an AllUsers one. In WSR-TC this will be spit out at the /VSIX/15Ins/ location. 

As the VS insertion depends on a <packageId>.json file that's generated next to a VSIX now, we propagate that to our NuGetCi drop. 

PS. Once this change is in, all private branches will need to rebase this change in, because of the multitude of CI changes.

//cc
@rrelyea @emgarten @alpaix @jainaashish @mishra14 @zhili1208 @rohit21agrawal

### CODE REVIEW

File: scripts/e2etests/InstallNuGetVSIX.ps1 [hunk 1]
```diff
@@ -33,7 +33,6 @@ $VSIXSrcPath = Join-Path $NuGetDropPath 'NuGet.Tools.vsix'
 $VSIXPath = Join-Path $FuncTestRoot 'NuGet.Tools.vsix'
 
 Copy-Item $VSIXSrcPath $VSIXPath
-
```
Comment (alpaix) [inline scripts/e2etests/InstallNuGetVSIX.ps1]: Revert if there's no actual changes #Resolved

File: src/NuGet.Clients/NuGet.Tools/NuGet.Tools.csproj [hunk 2]
```diff
@@ -30,7 +30,14 @@
   </PropertyGroup>
   <PropertyGroup Condition="'$(VisualStudioVersion)' == '15.0'">
     <VsixOutputDirName>VS15</VsixOutputDirName>
+    <IsProductComponent>true</IsProductComponent>
   </PropertyGroup>
+  <PropertyGroup Condition="'$(VisualStudioVersion)' == '15.0' AND '$(Insertable)' == 'true'">
```
Comment (alpaix) [inline src/NuGet.Clients/NuGet.Tools/NuGet.Tools.csproj]: Rename to `$(IsInsertable)` similar to `$(IsProductComponent)` #Resolved

File: src/NuGet.Clients/NuGet.Tools/NuGet.Tools.csproj [hunk 3]
```diff
@@ -334,13 +344,44 @@
...
+  </Target>
+  <Target Name="AfterBuild15" Condition="'$(VisualStudioVersion)' == '15.0'">
     <PropertyGroup>
       <PublishDestination>$(ArtifactRoot)\$(VsixOutputDirName)</PublishDestination>
     </PropertyGroup>
-    <Message Text="Publishing VSIX package files..." Importance="high" />
+    <Message Text="Publishing VS15 VSIX package files..." Importance="high" />
     <Message Text="$(TargetVsixContainer) -&gt; $(PublishDestination)" Importance="high" />
     <Copy SourceFiles="$(TargetVsixContainer)" DestinationFolder="$(PublishDestination)" OverwriteReadOnlyFiles="true" />
+    <Message Text="$(ArtifactRoot)\NuGet.Tools\15.0\$(Configuration)\Microsoft.VisualStudio.NuGet.Core.json -&gt; $(PublishDestination)" Importance="high" />
```
Comment (alpaix) [inline src/NuGet.Clients/NuGet.Tools/NuGet.Tools.csproj]: Define a property for the full path #Closed

File: src/NuGet.Clients/NuGet.Tools/source.extension.vs15.vsixmanifest [hunk 4]
```diff
@@ -2,13 +2,16 @@
 <PackageManifest Version="2.0.0" xmlns="http://schemas.microsoft.com/developer/vsx-schema/2011" xmlns:d="http://schemas.microsoft.com/developer/vsx-schema-design/2011">
   <Metadata>
     <Identity Id="NuGet.72c5d240-f742-48d4-a0f1-7016671e405b" Version="99.99.99" Language="en-US" Publisher="Microsoft Corporation" />
-    <DisplayName>NuGet Package Manager for Visual Studio 15</DisplayName>
+    <DisplayName>NuGet Package Manager for Visual Studio 17</DisplayName>
```
Comment (alpaix) [inline src/NuGet.Clients/NuGet.Tools/source.extension.vs15.vsixmanifest]: Visual Studio 2017 or Visual Studio "15" #Resolved

File: src/NuGet.Clients/NuGet.Tools/source.extension.vs15.insertable.vsixmanifest [hunk 5]
```diff
@@ -0,0 +1,39 @@
+<?xml version="1.0" encoding="utf-8"?>
+<PackageManifest Version="2.0.0" xmlns="http://schemas.microsoft.com/developer/vsx-schema/2011" xmlns:d="http://schemas.microsoft.com/developer/vsx-schema-design/2011">
+  <Metadata>
+    <Identity Id="NuGet.72c5d240-f742-48d4-a0f1-7016671e405b" Version="99.99.99" Language="en-US" Publisher="Microsoft Corporation" />
+    <DisplayName>NuGet Package Manager for Visual Studio 17</DisplayName>
```
Comment (alpaix) [inline src/NuGet.Clients/NuGet.Tools/source.extension.vs15.insertable.vsixmanifest]: Fix the name #Resolved

File: src/NuGet.Clients/NuGet.Tools/NuGet.Tools.csproj (hunk 3 above)
Comment (alpaix) [reply src/NuGet.Clients/NuGet.Tools/NuGet.Tools.csproj]: Also, use `$(OutDir)` for the output file location

File: src/NuGet.Clients/NuGet.Tools/NuGet.Tools.csproj [hunk 6]
```diff
@@ -334,13 +344,44 @@
     </None>
   </ItemGroup>
   <ItemGroup />
-  <Target Name="AfterBuild">
+  <Target Name="AfterBuild" DependsOnTargets="AfterBuild14;AfterBuild15;AfterBuildIns15"/>
```
Comment (alpaix) [inline src/NuGet.Clients/NuGet.Tools/NuGet.Tools.csproj]: I'd suggest naming targets respectively: `PublishVS14`, `PublishVS15`, and `PublishVS15Insertable` #Resolved

File: build.ps1 [hunk 7]
```diff
@@ -131,6 +131,19 @@ Invoke-BuildStep 'Building NuGet.Clients projects - VS15 Toolset' {
...
+Invoke-BuildStep 'Building NuGet.Tools.vsix for VS Insertion - VS15 Toolset' {
+        Build-ClientsProjectHelper `
+        -SolutionOrProject (Join-Path $NuGetClientRoot .\src\NuGet.Clients\NuGet.Tools\NuGet.Tools.csproj -Resolve)`
+        -Configuration $Configuration `
+        -ReleaseLabel $ReleaseLabel `
+        -BuildNumber $BuildNumber `
+        -Parameters @{'Insertable'='true'} `
+        -ToolsetVersion 15 `
+    } `
+    -skip:$SkipVS15 `
```
Comment (alpaix) [inline build.ps1]: I presume this build step should be running on CI only. If so then
`-skip:($SkipVS15 -or -not $CI)` #Resolved

File: src/NuGet.Clients/NuGet.Tools/NuGet.Tools.csproj (hunk 3 above)
Comment (alpaix) [reply src/NuGet.Clients/NuGet.Tools/NuGet.Tools.csproj]: Actually this file should be an item, similar to `$(NuGetSymbolFiles)` #Resolved

**Review by alpaix — CHANGES_REQUESTED** [review]
Address the comments.

File: src/NuGet.Clients/NuGet.Tools/NuGet.Tools.csproj (hunk 3 above)
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.Tools/NuGet.Tools.csproj]: Change PublishDestination to OutDir or? #Resolved

Comment (alpaix) [reply src/NuGet.Clients/NuGet.Tools/NuGet.Tools.csproj]: Path to the generated file should be `$(OutputPath)Microsoft.VisualStudio.NuGet.Core.json`

File: build.ps1 (hunk 7 above)
Comment (nkolev92) [inline build.ps1]: It should run on local builds as well. That way it's easy to install your own local bits.

**Comment by nkolev92** [comment]
🔔

File: src/NuGet.Clients/NuGet.Tools/NuGet.Tools.csproj [hunk 8]
```diff
@@ -334,13 +344,50 @@
...
+    <Message Text="Copying symbols file '%(NuGetSymbolFiles.Identity)'" Condition=" '@(NuGetSymbolFiles)' != '' " />
+    <Copy SourceFiles="@(NuGetSymbolFiles)" DestinationFolder="$(PublishDestination)\symbols" OverwriteReadOnlyFiles="true" />
+  </Target>
+  <Target Name="PublishVS15" Condition="'$(VisualStudioVersion)' == '15.0'">
     <PropertyGroup>
       <PublishDestination>$(ArtifactRoot)\$(VsixOutputDirName)</PublishDestination>
     </PropertyGroup>
-    <Message Text="Publishing VSIX package files..." Importance="high" />
+    <ItemGroup>
+      <ManifestPath Include="$(OutputPath)Microsoft.VisualStudio.NuGet.Core.json"/>
```
Comment (alpaix) [inline src/NuGet.Clients/NuGet.Tools/NuGet.Tools.csproj]: if it's an item (not property) it should be named like `ManifestV3File`

**Review by alpaix — APPROVED** [review]
Overall looks :sparkles:

Address the suggestion for better item naming.
Squash before merging!

**Pull request merged**